File: python/day1/main.py
import sys
import aoc


def part1(numbers):
    for n in numbers:
        other = 2020 - n
        if other in numbers:
            print(f"{n} * {other} = {n * other}")
            return


def part2(numbers):
    for n1 in numbers:
        for n2 in numbers:
            if n1 == n2:
                continue
            n3 = 2020 - n1 - n2
            if n3 in numbers:
                print(f"{n1} * {n2} * {n3} = {n1 * n2 * n3}")
                return

input = aoc.get_input('day1.txt')
numbers = {int(line) for line in input.splitlines()}
part1(numbers)
part2(numbers)

File: ruby/day1/main.rb
#!/usr/bin/env ruby

require 'set'
require_relative '../aoc/aoc'

s = AOC::Solution.new

s.preprocess {_1.split.map(&:to_i).to_set}

s.part1 do |numbers|
  numbers.each do |n|
    other = 2020 - n
    if numbers.include?(other)
      return n * other
    end
  end
end

s.part2 do |numbers|
  numbers.each do |n1|
    numbers.each do |n2|
      next if n1 == n2
      n3 = 2020 - n2 - n1
      if numbers.include?(n3)
        return n1 * n2 * n3
      end
    end
  end
end

s.exec(1)